cli/flox/src/commands/init/node.rs

```diff
@@ -69,38 +98,153 @@ enum NodeAction {
-            match (package_lock_exists, yarn_lock_exists) {
-                (false, false) => None,
-                (true, false) => Some(NodePackageManager::Npm),
-                (false, true) => Some(NodePackageManager::Yarn),
-                (true, true) => Some(NodePackageManager::Both),
-            }
+        // Get value for self.package_json_node_version if we didn't find a valid npm or yarn
+        let package_json_node_version = if package_manager.is_some() {
+            PackageJSONVersion::None
         } else {
+            match versions {
+                Some(PackageJSONVersions {
+                    node: Some(node_version),
+                    ..
+                }) => match Self::try_find_compatible_version("nodejs", &node_version, None, flox)?
+                {
+                    None => PackageJSONVersion::Unavailable,
+                    Some(result) => PackageJSONVersion::Found(Box::new(result)),
+                },
+                Some(_) => PackageJSONVersion::Unspecified,
+                _ => PackageJSONVersion::None,
+            }
+        };
+
+        // Get value for self.nvmrc_version if we didn't find a valid npm or yarn
+        let nvmrc_version = if package_manager.is_some() {
             None
+        } else {
+            match package_json_node_version {
+                // package.json is higher priority than .nvmrc,
+                // so don't check .nvmrc if we know we'll use the version in
+                // package.json or we know we can't provide it
+                PackageJSONVersion::Found(_) | PackageJSONVersion::Unavailable => None,
+                _ => Self::get_nvmrc_version(path, flox)?,
+            }
         };
 
         Ok(Self {
+            package_json_node_version,
             nvmrc_version,
-            package_json_version,
-            hook,
+            package_manager,
         })
     }
 
+    /// Look for nodejs, npm, and yarn versions in a (possibly non-existent)
+    /// `package.json` file
+    fn get_package_json_versions(path: &Path) -> Result<Option<PackageJSONVersions>> {
+        let package_json = path.join("package.json");
+        if !package_json.exists() {
+            return Ok(None);
+        }
+        let package_json_contents = fs::read_to_string(package_json)?;
+        match serde_json::from_str::<serde_json::Value>(&package_json_contents) {
+            // Treat a package.json that can't be parsed as JSON the same as it not existing
+            Err(_) => Ok(None),
+            Ok(package_json_json) => {
+                let node = package_json_json["engines"]["node"]
+                    .as_str()
+                    .map(|s| s.to_string());
+                let npm = package_json_json["engines"]["npm"]
+                    .as_str()
+                    .map(|s| s.to_string());
+                let yarn = package_json_json["engines"]["yarn"]
+                    .as_str()
+                    .map(|s| s.to_string());
```

Comment:
quib: node and npm are both "engines"? node is weird

Reply:
🤷